### PR TITLE
[RW-12857][risk=no] [API] Add APIs to support user publish community workspaces

### PR DESCRIPTION
**Pull request by NehaBroad**
https://github.com/user-attachments/assets/cf29dafc-6eb6-4924-9f66-789e0bcfa25a

---
**PR checklist**

- [ ] I have included an issue ID or "no ticket" in the PR title as outlined in [CONTRIBUTING.md](https://github.com/all-of-us/workbench/blob/main/.github/CONTRIBUTING.md).
- [ ] I have included a risk tag of the form `[risk=no|low|moderate|severe]` in the PR title as outlined in [CONTRIBUTING.md](https://github.com/all-of-us/workbench/blob/main/.github/CONTRIBUTING.md).
- [ ] I have manually tested this change and my testing process is described above.
- [ ] This change includes appropriate automated tests, and I have documented any behavior that cannot be tested with code.
- [ ] I have added explanatory comments where the logic is not obvious.
- One or more of the following is true:
  - [ ] This change is intended to complete a JIRA story, so I have checked that all AC are met for that story.
  - [ ] This change fixes a bug, so I have ensured the steps to reproduce are in the Jira ticket or provided above.
  - [ ] This change impacts deployment safety (e.g. removing/altering APIs which are in use), so I have documented the impacts in the description.
  - [ ] This change includes a new feature flag, so I have created and linked new JIRA tickets to (a) turn on the feature flag and (b) remove it later.
  - [ ] This change modifies the UI, so I have taken screenshots or recordings of the new behavior and notified the PO and UX designer in [Slack](https://pmi-engteam.slack.com/archives/C02MWP2RN5P).
  - [ ] This change modifies API behavior, so I have run the relevant E2E tests locally because API changes are not covered by our PR checks.
  - [ ] None of the above apply to this change.

### CODE REVIEW

File: api/src/main/java/org/pmiops/workbench/utils/mappers/FeaturedWorkspaceMapper.java [hunk 1]
```diff
@@ -16,6 +16,12 @@ public interface FeaturedWorkspaceMapper {
   DbFeaturedWorkspace toDbFeaturedWorkspace(
       PublishWorkspaceRequest featuredWorkspace, DbWorkspace dbWorkspace);
 
+  @Mapping(target = "workspace", source = "dbWorkspace")
+  @Mapping(target = "description", ignore = true)
+  @Mapping(target = "id", ignore = true)
+  DbFeaturedWorkspace toDbFeaturedWorkspace(
```
Comment (NehaBroad) [inline api/src/main/java/org/pmiops/workbench/utils/mappers/FeaturedWorkspaceMapper.java:22]: I'll have a clean up PR later to remove description field

File: api/src/main/java/org/pmiops/workbench/mail/MailServiceImpl.java [hunk 2]
```diff
@@ -388,23 +387,33 @@ public void sendFileLengthsEgressRemediationEmail(DbUser dbUser, EgressRemediati
   }
 
   @Override
-  public void sendPublishWorkspaceByAdminEmail(
+  public void sendPublishWorkspaceEmail(
```
Comment (NehaBroad) [inline api/src/main/java/org/pmiops/workbench/mail/MailServiceImpl.java:390]: Support has requested that we do not send an email notification when an admin publishes a workspace, as these are managed by the DRC. I have additional questions about this process, so for now, I am combining the two publish methods (by Admin and by Owner).

Depending on the answers I receive, I'll create a new PR to remove the email call from the admin service.

File: api/src/main/java/org/pmiops/workbench/firecloud/FireCloudServiceImpl.java [hunk 3]
```diff
@@ -386,6 +387,26 @@ public RawlsWorkspaceACLUpdateResponseList updateWorkspaceACL(
...
+   * access to all users in the groupEmail If Unpublished: Grant No access to all users in the
+   * groupEmail
+   *
+   * @param workspaceNamespace the Namespace (Terra Billing Project) of the Workspace to modify
+   * @param firecloudName the Terra Name of the Workspace to modify
+   * @param groupEmail the email of the group to update
+   * @return
+   */
+  @Override
+  public void updatePublishWorkspaceACL(
```
Comment (jmthibault79) [inline api/src/main/java/org/pmiops/workbench/firecloud/FireCloudServiceImpl.java:401]: something feels not quite right about the name here.  But it's close.  how about `updateWorkspaceAclForPublishing()` ?

File: api/src/main/java/org/pmiops/workbench/firecloud/FireCloudServiceImpl.java [hunk 4]
```diff
@@ -386,6 +387,26 @@ public RawlsWorkspaceACLUpdateResponseList updateWorkspaceACL(
...
+      String workspaceNamespace, String firecloudName, String groupEmail, boolean publish) {
```
Comment (jmthibault79) [inline api/src/main/java/org/pmiops/workbench/firecloud/FireCloudServiceImpl.java:402]: Since we want to tightly restrict `groupEmail` to the same value (the RT auth domain group), would do you think about calling `workspaceService.getPublishedWorkspacesGroupEmail()` inside this method instead?

File: api/src/main/java/org/pmiops/workbench/workspaceadmin/WorkspaceAdminServiceImpl.java [hunk 5]
```diff
@@ -546,27 +555,6 @@ private void sendEmailToWorkspaceOwners(
     }
   }
 
-  private void updateACLForFeaturedWorkspace(
```
Comment (jmthibault79) [inline api/src/main/java/org/pmiops/workbench/workspaceadmin/WorkspaceAdminServiceImpl.java:549]: makes sense to move these - good consolidation/reuse.

File: api/src/main/java/org/pmiops/workbench/workspaces/WorkspaceServiceImpl.java [hunk 6]
```diff
@@ -495,4 +515,47 @@ public void updateFreeTierWorkspacesStatus(final DbUser user, final BillingStatu
...
   }
+
+  @Override
+  public void publishCommunityWorkspace(DbWorkspace dbWorkspace) {
+    featuredWorkspaceDao
+        .findByWorkspace(dbWorkspace)
+        .ifPresentOrElse(
+            (dbFeaturedWorkspace) -> {
+              // If the workspace is already featured, a BadRequestException is thrown
+              throw new BadRequestException("Workspace is already marked featured");
```
Comment (jmthibault79) [inline api/src/main/java/org/pmiops/workbench/workspaces/WorkspaceServiceImpl.java:526]: "already published" ?

File: api/src/test/java/org/pmiops/workbench/api/WorkspacesControllerTest.java [hunk 7]
```diff
@@ -3014,4 +3011,32 @@ public void getWorkspaceResources() {
...
+  public void testPublishCommunityWorkspace_ByUserWithWriterAccess() {
+    assertThrows(
+        ForbiddenException.class,
+        () -> {
+          Workspace ws = createWorkspace();
+          ws = workspacesController.createWorkspace(ws).getBody();
+          ws.setName("updated-name");
+          stubGetWorkspace(
+              ws.getNamespace(), ws.getId(), ws.getCreator(), WorkspaceAccessLevel.WRITER);
+          workspacesController.publishCommunityWorkspace(ws.getNamespace());
```
Comment (jmthibault79) [inline api/src/test/java/org/pmiops/workbench/api/WorkspacesControllerTest.java:3025]: nit: I prefer the style where `assertThrows()` only runs one command, so it's clearer which one throws.  Like this:
```
assertThrows(
        ForbiddenException.class,
        () -> workspacesController.publishCommunityWorkspace(ws.getNamespace()));
```

File: api/src/test/java/org/pmiops/workbench/api/WorkspacesControllerTest.java [hunk 8]
```diff
@@ -3014,4 +3011,32 @@ public void getWorkspaceResources() {
...
+        });
+  }
+
+  @Test
+  public void testPublishCommunityWorkspace_ByUserWithReaderAccess() {
```
Comment (jmthibault79) [inline api/src/test/java/org/pmiops/workbench/api/WorkspacesControllerTest.java:3030]: we should have a success case too, don't you think? :)

File: api/src/test/java/org/pmiops/workbench/workspaces/WorkspaceServiceTest.java [hunk 9]
```diff
@@ -128,18 +137,20 @@ DbUser user() {
     }
   }
 
+  @Autowired private AccessTierDao accessTierDao;
+  @Autowired private CdrVersionDao cdrVersionDao;
+  @Autowired private WorkspaceDao workspaceDao;
+  @Autowired private WorkspaceService workspaceService;
+
+  @MockBean private AccessTierService accessTierService;
```
Comment (jmthibault79) [inline api/src/test/java/org/pmiops/workbench/workspaces/WorkspaceServiceTest.java:145]: would you mind renaming this one to "mock" ?

File: api/src/test/java/org/pmiops/workbench/workspaces/WorkspaceServiceTest.java [hunk 10]
```diff
@@ -464,7 +475,9 @@ public void enforceFirecloudAclsInRecentWorkspaces() {
             RawlsWorkspaceAccessLevel.NO_ACCESS,
             WorkspaceActiveStatus.ACTIVE);
     workspaceService.updateRecentWorkspaces(sharedWorkspace);
-
+    when(mockWorkspaceAuthService.enforceWorkspaceAccessLevel(
```
Comment (jmthibault79) [inline api/src/test/java/org/pmiops/workbench/workspaces/WorkspaceServiceTest.java:478]: could you say a little about why this change is needed?

File: api/src/test/java/org/pmiops/workbench/workspaces/WorkspaceServiceTest.java [hunk 11]
```diff
@@ -677,6 +690,79 @@ public void testGetWorkspace_featuredCategory() {
...
+        .thenReturn(Optional.of(mockDBFeaturedWorkspace));
+
+    // Assert
+    assertThrows(
+        BadRequestException.class,
+        () -> workspaceService.publishCommunityWorkspace(dbWorkspace),
+        "Workspace is already featured");
+  }
+
+  public static DbAccessTier createRegisteredTier() {
```
Comment (jmthibault79) [inline api/src/test/java/org/pmiops/workbench/workspaces/WorkspaceServiceTest.java:718]: This should be removed - it duplicates the same code in `TestMockFactory`

**Review by jmthibault79 — APPROVED** [review]
Works!  Some comments but nothing huge.

File: api/src/main/java/org/pmiops/workbench/firecloud/FireCloudServiceImpl.java (hunk 4 above)
Comment (NehaBroad) [reply api/src/main/java/org/pmiops/workbench/firecloud/FireCloudServiceImpl.java:402]: I wanted to do this, but since the Workspace Service is also calling it, I was concerned about creating a circular reference. I could call accessTierService.getRegisteredTierOrThrow().getAuthDomainGroupEmail() within this method, but that would result in duplicate code. I'll consider if the  method  getPublishedWorkspacesGroupEmail can be moved to another location to avoid this issue.

File: api/src/test/java/org/pmiops/workbench/workspaces/WorkspaceServiceTest.java (hunk 10 above)
Comment (NehaBroad) [reply api/src/test/java/org/pmiops/workbench/workspaces/WorkspaceServiceTest.java:478]: I moved class workspaceAuth from import to mockBean i'll try to add meaningful comment

File: api/src/main/java/org/pmiops/workbench/firecloud/FireCloudServiceImpl.java (hunk 4 above)
Comment (jmthibault79) [reply api/src/main/java/org/pmiops/workbench/firecloud/FireCloudServiceImpl.java:402]: One think we can do is pass `workspaceService` as the parameter.  When it's WorkspaceService itself calling the method, it can pass `this`.

File: api/src/main/java/org/pmiops/workbench/mail/MailServiceImpl.java [hunk 12]
```diff
@@ -388,23 +387,33 @@ public void sendFileLengthsEgressRemediationEmail(DbUser dbUser, EgressRemediati
...
       DbWorkspace workspace, List<DbUser> owners, FeaturedWorkspaceCategory publishCategory)
       throws MessagingException {
-    sendPublishUnpublishWorkspaceByAdminEmail(
-        workspace, owners, featuredWorkspaceCategoryAsDisplayString(publishCategory), true);
+
+    final String ownersForLogging =
+        owners.stream().map(this::userForLogging).collect(Collectors.joining(", "));
```
[the remark this answers is not in the document]
Comment (jmthibault79) [reply api/src/main/java/org/pmiops/workbench/mail/MailServiceImpl.java]: please use the function instead

File: api/src/main/java/org/pmiops/workbench/firecloud/FireCloudServiceImpl.java [hunk 13]
```diff
@@ -157,6 +161,8 @@ public FireCloudServiceImpl(
       RawlsRetryHandler rawlsRetryHandler,
       CalhounRetryHandler calhounRetryHandler,
       SamRetryHandler samRetryHandler,
+      // To avoid circular dependencies, we use @Lazy to defer the creation of WorkspaceService
```
Comment (jmthibault79) [inline api/src/main/java/org/pmiops/workbench/firecloud/FireCloudServiceImpl.java:164]: good find!  This may be useful elsewhere too

File: api/src/test/java/org/pmiops/workbench/workspaces/WorkspaceServiceTest.java (hunk 10 above)
Comment (jmthibault79) [reply api/src/test/java/org/pmiops/workbench/workspaces/WorkspaceServiceTest.java:478]: ok cool

File: api/src/test/java/org/pmiops/workbench/api/WorkspacesControllerTest.java [hunk 14]
```diff
@@ -3014,4 +3019,36 @@ public void getWorkspaceResources() {
...
+        () -> {
+          workspacesController.publishCommunityWorkspace(wsNamespace);
+        });
+  }
+
+  @Test
+  public void testPublishCommunityWorkspace() {
+    Workspace ws = createWorkspaceAndGrantAccess(WorkspaceAccessLevel.OWNER);
+    String wsNamespace = ws.getNamespace();
+    doNothing().when(workspaceService).publishCommunityWorkspace(any(DbWorkspace.class));
```
Comment (jmthibault79) [inline api/src/test/java/org/pmiops/workbench/api/WorkspacesControllerTest.java:3049]: is this because it's a real impementation and not a mock?

File: api/src/test/java/org/pmiops/workbench/workspaces/WorkspaceServiceTest.java [hunk 15]
```diff
@@ -677,6 +690,68 @@ public void testGetWorkspace_featuredCategory() {
...
+            .setCategory(DbFeaturedWorkspace.DbFeaturedCategory.TUTORIAL_WORKSPACES);
+
+    when(mockFeaturedWorkspaceDao.findByWorkspace(dbWorkspace))
+        .thenReturn(Optional.of(mockDBFeaturedWorkspace));
+
+    // Assert
+    assertThrows(
+        BadRequestException.class,
+        () -> workspaceService.publishCommunityWorkspace(dbWorkspace),
+        "Workspace is already featured");
```
Comment (jmthibault79) [inline api/src/test/java/org/pmiops/workbench/workspaces/WorkspaceServiceTest.java:715]: shouldn't this be "published" ?

Comment (NehaBroad) [reply api/src/test/java/org/pmiops/workbench/workspaces/WorkspaceServiceTest.java:715]: Why is this passing, i'll check

Comment (NehaBroad) [reply api/src/test/java/org/pmiops/workbench/workspaces/WorkspaceServiceTest.java]: assertThrows doesnt confirm the message for exception as 

 _@Rule and ExpectedException are part of JUnit 4, but when you annotate your test with @org.junit.jupiter.api.Test then the test will be executed by JUnit 5 and JUnit 5 doesn't know how to process @Rule annotations_
 
 Workaround to get the exception object and then check the message
